sdk/opa.go

```diff
@@ -298,6 +298,96 @@ func newDecisionResult() (*DecisionResult, error) {
 	return result, nil
 }
 
+// Partial returns a named decision. This function is threadsafe.
+func (opa *OPA) Partial(ctx context.Context, options PartialOptions) (*DecisionResult, error) {
+
+	if options.Resolver == nil {
+		return nil, fmt.Errorf("ResultResolver must be specified")
+	}
+
+	m := metrics.New()
+	m.Timer(metrics.SDKDecisionEval).Start()
+
+	result, err := newDecisionResult()
+	if err != nil {
+		return nil, err
+	}
+
+	opa.mtx.Lock()
+	s := *opa.state
+	opa.mtx.Unlock()
+
+	record := server.Info{
+		DecisionID: result.ID,
+		Timestamp:  options.Now,
+		Input:      &options.Input,
+		Query:      options.Query,
+		Path:       options.Path,
+		Metrics:    m,
+	}
+
+	if record.Timestamp.IsZero() {
+		record.Timestamp = time.Now().UTC()
+	}
+
+	if record.Path == "" {
+		record.Path = *s.manager.Config.DefaultDecision
+	}
+
+	record.Txn, record.Error = s.manager.Store.NewTransaction(ctx, storage.TransactionParams{})
+
+	if record.Error == nil {
+		defer s.manager.Store.Abort(ctx, record.Txn)
+		var pq *rego.PartialQueries
+		pq, record.InputAST, record.Bundles, record.Error = partial(ctx, partialEvalArgs{
+			runtime:   s.manager.Info,
+			printHook: s.manager.PrintHook(),
+			compiler:  s.manager.GetCompiler(),
+			store:     s.manager.Store,
+			txn:       record.Txn,
+			now:       record.Timestamp,
+			query:     record.Query,
+			unknowns:  options.Unknowns,
+			input:     *record.Input,
+			m:         record.Metrics,
+		})
+		if record.Error == nil {
+			var recordResults interface{}
+			result.Result, recordResults, err = options.Resolver.ProcessResults(pq)
+			if err != nil {
+				return nil, err
+			}
+			record.Results = &recordResults
+		}
+	}
+
+	m.Timer(metrics.SDKDecisionEval).Stop()
+
+	if logger := logs.Lookup(s.manager); logger != nil {
+		if err := logger.Log(ctx, &record); err != nil {
+			return result, fmt.Errorf("decision log: %w", err)
+		}
+	}
+
+	return result, record.Error
+}
+
+// The first interface being returned is the type that will be used for further processing
+// The second interface is for the decision logger, most of the time these should be the same
+type ResultResolver interface {
+	ProcessResults(pq *rego.PartialQueries) (interface{}, interface{}, error)
+}
```

Comment:
I'm a bit uncertain about this interface because the system consuming the decison logs won't necessarily have access to the implementation so it won't be able to replay the policy decision. On the other hand, I could see how mapping the partial eval results is valuable.

Reply:
As I started coding this up, it became apparent that not having the mapped results being logged the decisions became very difficult to understand.  Mapping them to the context that the developer is more likely to understand, the decisions become far more valuable.

I have to admit I didn't think about the downstream affect that this could have on systems built to replay decisions.  Maybe having contrib repos setup with the most common mappers would help these systems pull them in and support it better.  (I had originally thought of including additional mappers in the SDK, but did not want to muddy up the OPA codebase)